### DISCUSSION TITLE
Why are my Morphia Cursors stacking up ?

### DISCUSSION THREAD

**Discussion by tariqanas** [MorphiaOrg/morphia#3114] [General]
Hello,

I work in a large project , for migrating Morphia from 1.6 to 2.4.11 .
I updated all the DAO to the new versions, but while looking at Grafana , i see that the number of cursors used by this version is greater than the number used by morphia 1.6
I understand that in morphia 1.6 the cursors were handled internally, which is not the case in 2.4.11

Is it a normal behaviour ? we've been analysing this change of numbers but couldn't find an explanation. Is it a normal behaviour ?
We put all the morphia cursors inside a try-with-resource, but we keep seeing them stacking up

Any help or guidance would be much appreciated.

    Server Version: 5.0
    Driver Version: 4.11
    Morphia Version: 2.4

**Comment by evanchooly** [comment]
I'm not sure i'm following the cursors being handled internally but what do you mean when you cursors stacking up?  Do you mean connections to the database?

**Reply by tariqanas** [reply]
I mean by that, the MorphiaCursors  process the documents, but take so much time to close . which make cursors stack up everytime we have new calls to the api. knowing that it's just a find .
The code is just something like that (Sorry can't share the real code )  .



 ```java
 FindOptions findOptions = new FindOptions()
                                      .limit(10)
                                      .sort(Sort.ascending("fieldName")); 
Query<YourEntity> query = morphia.find(YourEntity.class);

// Execute the query with FindOptions
List<YourEntity> results = query.iterator(findOptions).toList();

for (YourEntity entity : results) {
    //Do_stuff.
}
```

I wanted to know if there was a difference between how Morphia handles cursors between Morphia 1.6 & 2.4. We can't explain what's happening. 

Thank you

**Reply by evanchooly** [reply]
well, 2.0 was a pretty drastic architectural upgrade so there are loads of changes internally for sure.  As far as the cursors go, by the time `toList()` returns Morphia will have already called `close()` on the cursor so it's up to the driver to manage things from that point.

**Reply by tariqanas** [reply]
Ok , thank you i'll keep investigating. 
To this day , i switched query.iterator(findOptions) to query.stream(findOptions). But still same . 
Maybe i should explore other things rather than Morphia.  
The Java Mongo driver version seems Ok to you  to use with Morphia 2.4.11 ?  
Or maybe we should try higher morphia versions like the last one ? 2.4.14 ?

If you don't see anything that can help us , related to Morphia . You may close the discussion .

Thank you V.much Evanchooly.

**Reply by evanchooly** [reply]
nothing obvious in morphia comes to mind but I don't have a lot of detail to work.  there's not really a difference between `.11` and `.14` so far as this is concerned.  If I had to guess, i would say it's likely latency on closing closing sockets or responsiveness on the server responding to cursor closing.  Hard to tell from here, though.

**Reply by tariqanas** [reply]
Hello , 
Just wanted to say that i found  what caused the issue . 
I had a large query that took time to fetch the data  & since it was called many times, with others queries the cursors stacked up on earch others & looked scary .

Beside , i put the cursors in a try with resources  to be sure that they were automatically managed.

Things look fine now . 
Thank you

**Reply by evanchooly** [reply]
Oh that's great news.  Nice work tracking that down.
